control-plane/agents/src/bin/core/volume/specs.rs

```diff
@@ -173,6 +181,84 @@ pub(crate) async fn nexus_attach_candidates(
     Ok(candidates)
 }
 
+pub(crate) struct CreateReplicas {
+    replicas: Vec<CreateReplica>,
+    affinity_labels: Option<HashMap<String, String>>,
+}
+
+impl CreateReplicas {
+    /// Create a new `CreateReplicas` with replicas and optional labels.
+    pub fn new(
+        replicas: Vec<CreateReplica>,
+        affinity_labels: Option<HashMap<String, String>>,
+    ) -> Self {
+        Self {
+            replicas,
+            affinity_labels,
+        }
+    }
+    /// Get the replicas.
+    pub fn replicas(&self) -> &Vec<CreateReplica> {
+        &self.replicas
+    }
+
+    /// Get the labels.
+    pub fn labels(&self) -> Option<HashMap<String, String>> {
+        self.affinity_labels.clone()
+    }
+}
+
+/// Return a list of appropriate requests which can be used to create a replica on a pool.
+/// This can be used when the volume's current replica count is smaller than the desired volume's
+/// replica count.
+pub(crate) async fn volume_replica_candidates_new(
+    registry: &Registry,
+    volume_spec: &VolumeSpec,
+) -> Result<Vec<CreateReplicas>, SvcError> {
+    let request = GetSuitablePools::new(volume_spec, None);
+    let mut vec_create_replicas: Vec<CreateReplicas> = Vec::new();
+    let set_of_pools: Vec<Vec<crate::controller::wrapper::PoolWrapper>> =
+        scheduling::volume_pool_candidates_new(request.clone(), registry).await;
+
+    for pools in set_of_pools.iter() {
+        if pools.is_empty() {
+            return Err(SvcError::NotEnoughResources {
+                source: NotEnough::OfPools { have: 0, need: 1 },
+            });
+        }
+        volume_spec.trace(&format!(
+            "Creation pool candidates for volume: {:?}",
+            pools.iter().map(|p| p.state()).collect::<Vec<_>>()
+        ));
+        let mut create_replicas: Vec<CreateReplica> = Vec::new();
+        let mut labels = HashMap::new();
+        for pool_wrapper in pools.iter() {
+            let replica_uuid = ReplicaId::new();
+            let create_replica = CreateReplica {
+                node: pool_wrapper.node.clone(),
```

Comment:
hmm isn't there already code which setups the CreateReplica?

Reply:
its there but its in `volume_replica_candidates` which uses `volume_pool_candidates` . But in this case we are using `volume_pool_candidates_new`

Reply:
Can we break down the existing function so we can reuse setting up the creation of the Vec<CreateReplica> ?

Reply:
We cant as for every pool we are setting the labels as well .

Reply:
hmm I'm sure it can still be done somehow, but that's fine, I can take this up in refactoring later